### ISSUE TITLE
bug:  Potential memory leak on the status-addon agent

### ISSUE THREAD

**Issue by dumb0002** [kubestellar/kubestellar#2040]
### Describe the bug

Conducted a long duration experiment running using KS-v0.21.2 to check for any resource leaks: in this experiment I **created** and **deleted** 68,436 pods. One pod is created in 5 seconds and at most we had 4 pods running in parallel in the system - This experiment run for **4 days** and each pod is deleted after successful execution.

The plot below shows that memory of the ****status-addon agent** increases **linearly** with the number of objects created in the WEC while workloads objects are being deleted after successful execution: I also show similar behavior for the status agent for ks-v0.20 on the long duration experiment:

<img width="941" alt="Screenshot 2024-04-09 at 3 14 53 PM" src="https://github.com/kubestellar/kubestellar/assets/25727844/7ceb89fc-9773-4392-9b29-aae1e5927be3">

**Potential root-cause:** the status addon agent is keeping state for already deleted objects from the cluster. I constantly see the following messages in the logs for the status agent: 

```
2024-04-09T12:45:04Z	INFO	object in work queue no longer exists	{"key": "openshift-operator-lifecycle-manager/collect-profiles-28544400-qhmbt", "lister key": "v1/Pod"}
2024-04-09T12:48:23Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-n4b76", "lister key": "v1/Pod"}
2024-04-09T12:52:54Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-d2stm", "lister key": "v1/Pod"}
2024-04-09T12:52:55Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-wmcrh", "lister key": "v1/Pod"}
2024-04-09T12:54:12Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-rf899", "lister key": "v1/Pod"}
2024-04-09T12:59:52Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-2gzrz", "lister key": "v1/Pod"}
2024-04-09T13:00:04Z	INFO	object in work queue no longer exists	{"key": "openshift-operator-lifecycle-manager/collect-profiles-28544415-wglfd", "lister key": "v1/Pod"}
2024-04-09T13:03:51Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-v8q5f", "lister key": "v1/Pod"}
2024-04-09T13:05:06Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-8zv87", "lister key": "v1/Pod"}
2024-04-09T13:05:07Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-xvxgj", "lister key": "v1/Pod"}
2024-04-09T13:11:17Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-cngvj", "lister key": "v1/Pod"}
2024-04-09T13:15:06Z	INFO	object in work queue no longer exists	{"key": "openshift-operator-lifecycle-manager/collect-profiles-28544430-6pnp9", "lister key": "v1/Pod"}
2024-04-09T13:15:12Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-2skdf", "lister key": "v1/Pod"}
2024-04-09T13:15:17Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-wzhbk", "lister key": "v1/Pod"}
2024-04-09T13:15:25Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-b6q5h", "lister key": "v1/Pod"}
2024-04-09T13:23:28Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-xhgw9", "lister key": "v1/Pod"}
2024-04-09T13:25:24Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-926j2", "lister key": "v1/Pod"}
2024-04-09T13:26:21Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-7qjh7", "lister key": "v1/Pod"}
2024-04-09T13:26:49Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-8hr9z", "lister key": "v1/Pod"}
2024-04-09T13:30:04Z	INFO	object in work queue no longer exists	{"key": "openshift-operator-lifecycle-manager/collect-profiles-28544445-fjzzr", "lister key": "v1/Pod"}
2024-04-09T13:33:43Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-rkm45", "lister key": "v1/Pod"}
2024-04-09T13:36:39Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-wm2l4", "lister key": "v1/Pod"}
2024-04-09T13:37:04Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-x5f8b", "lister key": "v1/Pod"}
2024-04-09T13:37:04Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-qdhf2", "lister key": "v1/Pod"}
2024-04-09T13:45:04Z	INFO	object in work queue no longer exists	{"key": "openshift-operator-lifecycle-manager/collect-profiles-28544460-5krtg", "lister key": "v1/Pod"}
2024-04-09T13:46:46Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-operators-x8wzm", "lister key": "v1/Pod"}
2024-04-09T13:47:14Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/certified-operators-qv8s6", "lister key": "v1/Pod"}
2024-04-09T13:47:33Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/redhat-marketplace-jb9xr", "lister key": "v1/Pod"}
2024-04-09T13:48:14Z	INFO	object in work queue no longer exists	{"key": "openshift-marketplace/community-operators-8zzt8", "lister key": "v1/Pod"}
```

### Steps To Reproduce

Run the stress script described here with the parameters described below: https://github.com/kubestellar/kubestellar/tree/main/test/performance

**Workload Generation Function**:
  - clusterloader tuningSet RSteppedLoad with burstSize=1 and stepDelay=5 sec

**Exp workload**: https://github.com/dumb0002/perf-tests/blob/ks-provider/clusterloader2/testing/load/test-workload.yaml
    - Pod that sleeps for 20 seconds

**Execution:** run this experiment for more than 24 hours to see the behavior described above. 

### Expected Behavior

The memory status-addon agent should not increase linearly when workloads objects are being deleted in the WEC

### Additional Context

Here are the collected logs for the [status-agent-logs.zip](https://github.com/kubestellar/kubestellar/files/14923038/status-agent-logs.zip)
     a) **Collected on 04/08**: status-agent-ks-0.21.2-04082024.txt
     b) **Collected on 04/09:**  status-agent-ks-0.21.2-04092024.txt

**Comment by ezrasilvera** [comment]
While performance is low priority at this point, putting it on medium priority as it seems there is no boundary to the increase.

**Comment by effi-ofer** [comment]
@pdettori can I take this over?

**Comment by pdettori** [comment]
@effi-ofer yes, that would be very helpful, thank you!

**Comment by effi-ofer** [comment]
I just tried to repro this on my system.  I didn't had the metrics server setup so I couldn't track the memory in sufficient graduality.  But I did notice that that the binding.control object and the correspounding manifestwork object are not deleted when the workload payload becomes nil.  The wds and wec deployment objects have been deleted.  
```
effi@fortress:~/edge-mc/test/e2e/ginkgo$ kubectl --context wds1 get binding.control -A
NAME              AGE
nginx             25m
nginx-singleton   24m
effi@fortress:~/edge-mc/test/e2e/ginkgo$ kubectl --context wds1 get binding.control nginx-singleton -o yaml
apiVersion: control.kubestellar.io/v1alpha1
kind: Binding
metadata:
  creationTimestamp: "2024-04-22T12:31:15Z"
  finalizers:
  - transport.kubestellar.io/object-cleanup
  generation: 2
  name: nginx-singleton
  ownerReferences:
  - apiVersion: control.kubestellar.io/v1alpha1
    blockOwnerDeletion: false
    controller: true
    kind: BindingPolicy
    name: nginx-singleton
    uid: c5824514-525e-4311-8514-f03f8e1475a7
  resourceVersion: "31377"
  uid: d32e0350-c7fa-41ac-949b-6097e07c8b86
spec:
  workload: {}
effi@fortress:~/edge-mc/test/e2e/ginkgo$ kubectl --context its1 get manifestworks -A
NAMESPACE   NAME                          AGE
cluster1    addon-addon-status-deploy-0   18h
cluster2    addon-addon-status-deploy-0   18h
cluster1    nginx-wds1                    25m
cluster2    nginx-wds1                    25m
```

**Comment by pdettori** [comment]
@effi-ofer I remember observing something similar. I would have expected the manifest work to be deleted when there is no workload  - that is, there should not be an empty manifestwork

**Comment by effi-ofer** [comment]
Sorry I wasn't clear.  The manifestwork object is deleted when empty.  The binding control object remains.   I see the binding object is deleted when the bindingpolicy is deleted.  I expect that's by design and if we ever feel the need to change this, we will.  I am going back to concentrating on the memory leak.

**Comment by effi-ofer** [comment]
Running the `ginkgo -v -focus "sets singleton status when a singleton bindingpolicy/deployment is created" -- -skip-setup` test in a loop I can see the memory used by wec1 (aka cluster1 in the test case) is slowly increasing.  After 300 iterations the memory foot print went up from 2.7GB to 3.0GB.  But I am using kind for my clusters and docker stats to track memory usage and this is not the most reliable way to gauge usage.   I did the same to wec2 and to the hub and I noticed the memory foot print increasing in a similar way as well.  

@dumb0002 you reproduced this on a real cluster?  Which one?
